lib/pure/asyncdispatch.nim

```diff
@@ -262,24 +277,121 @@ when defined(windows) or defined(nimdoc):
     PostCallbackDataPtr = ptr PostCallbackData
 
     AsyncEventImpl = object
-      hEvent: Handle
-      hWaiter: Handle
-      pcd: PostCallbackDataPtr
+      triggered: bool
+      when compileOption("threads"):
+        eventLock: RLock
+      p: PDispatcher
+      vFD: VirtualFD
+      callbacks: seq[Callback]
     AsyncEvent* = ptr AsyncEventImpl
 
     Callback = proc (fd: AsyncFD): bool {.closure,gcsafe.}
 
   proc hash(x: AsyncFD): Hash {.borrow.}
   proc `==`*(x: AsyncFD, y: AsyncFD): bool {.borrow.}
+  proc hash(x: VirtualFD): Hash {.borrow.}
+  proc `==`*(x: VirtualFD, y: VirtualFD): bool {.borrow.}
+
+  const InvalidVirtualFD = VirtualFD(-1)
+
+  {.push stackTrace:off.}
+  proc waitableCallback(param: pointer,
+                        timerOrWaitFired: WINBOOL): void {.stdcall.} =
+    var p = cast[PostCallbackDataPtr](param)
+    discard postQueuedCompletionStatus(p.ioPort, timerOrWaitFired.DWORD,
+                                       ULONG_PTR(p.handleFd),
+                                       cast[pointer](p.ovl))
+  {.pop.}
+
+  template registerWaitableHandle(p, hEvent, flags, pcd, timeout,
```

Comment:
Could this be a `proc` instead?

Reply:
IDK. I suppose so? It was this way when I found it. I saw no reason to change it. I just moved the template up because I needed it earlier in the code and Nim doesn't allow forward declarations.

Reply:
ahh, I didn't realise you moved it up. It really needs to be refactored IMO...